lib/devices_test.go

```diff
@@ -0,0 +1,197 @@
+package usb_tree_lib
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/google/gousb"
+)
+
+func TestDescToDevice(t *testing.T) {
+	desc := mockDesc()
+	dev := descToDevice(desc)
+	if dev.Name != "3.0 root hub (Linux Foundation)" {
+		t.Errorf("descToDevice failed: Name = %s, 3.0 root hub (Linux Foundation)", dev.Name)
+	}
+	if dev.VendorId != "1d6b" {
+		t.Errorf("descToDevice failed: VendorId = %s, want 1d6b", dev.VendorId)
+	}
+	if dev.ProductId != "0003" {
+		t.Errorf("descToDevice failed: ProductId = %s, want 0003", dev.ProductId)
+	}
+	if dev.Speed != "high" {
+		t.Errorf("descToDevice failed: Speed = %s, want high", dev.Speed)
+	}
+	if dev.Bus != 1 {
+		t.Errorf("descToDevice failed: Bus = %d, want 1", dev.Bus)
+	}
+}
+
+func TestDeviceDiff_Add(t *testing.T) {
+	d1 := Device{Path: []int{1}, Name: "A", VendorId: "v1", ProductId: "p1", Speed: "High", Bus: 1, State: StateNormal}
+	d2 := Device{Path: []int{2}, Name: "B", VendorId: "v2", ProductId: "p2", Speed: "High", Bus: 1, State: StateNormal}
+	d3 := Device{Path: []int{3}, Name: "C", VendorId: "v3", ProductId: "p3", Speed: "High", Bus: 1, State: StateNormal}
+	mockRefresh([]Device{d1, d2})
+	changed, merged := deviceDiff([]Device{d1, d2, d3})
+	if len(merged) != 3 {
+		t.Errorf("deviceDiff add failed: expected 3 devices, got %d", len(merged))
+	}
+	if !changed {
+		t.Errorf("deviceDiff add failed: expected changed=true, got false")
+	}
+	if !hasState(merged, StateAdded) {
+		t.Errorf("deviceDiff add failed: expected StateAdded in merged devices")
+	}
+}
+
+func TestDeviceDiff_Remove(t *testing.T) {
+	d1 := Device{Path: []int{1}, Name: "A", VendorId: "v1", ProductId: "p1", Speed: "High", Bus: 1, State: StateNormal}
+	d2 := Device{Path: []int{2}, Name: "B", VendorId: "v2", ProductId: "p2", Speed: "High", Bus: 1, State: StateNormal}
+	d3 := Device{Path: []int{3}, Name: "C", VendorId: "v3", ProductId: "p3", Speed: "High", Bus: 1, State: StateNormal}
+	mockRefresh([]Device{d1, d2, d3})
```

Comment:
You're now repeatedly crafting these lists in each test and it's hard to follow, I would create the devices on the root level with with names like `device1`, `device2`, ... and then create a couple of obviously named lists like `allDevices`, `partialDevices1`, `partialDevices2`, and so on that you can easily reuse

Reply:
Also could make the values for name, vid and pid slightly nicer
- `Device 1`, `0001`, `0010`
- `Device 2`, `0002`, `0020`
...